File: 94-binary-tree-inorder-traversal/binary-tree-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
    array := []int{}
    stack := []*TreeNode{}
    current := root

    for current != nil || len(stack) != 0 {
        for current != nil {
            stack = append(stack, current)
            current = current.Left
        }

        current = stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        array = append(array, current.Val)

        current = current.Right
    }

    return array
}

File: 94-binary-tree-inorder-traversal/binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        std::vector<int> array{};
        std::stack<TreeNode*> stack_{};
        TreeNode* current = root;

        while (current != nullptr || !stack_.empty()) {
            while (current != nullptr) {
                stack_.push(current);
                current = current->left;
            }

            current = stack_.top();
            stack_.pop();

            array.push_back(current->val);
            
            current = current->right;
        }

        return array;
    }
};